### PR TITLE
Revert ability for move and show/hide toolbars

### LINKED ISSUE #8680: Toolbar position is not saved or restored
## Current behaviour

Since #8572, I [always](https://github.com/gitextensions/gitextensions/pull/8572#issuecomment-734340761) end up with this toolbar at startup:
image

Even if I move the toolbars. Even when only on instance is opened.

Even with #8679 fix (to only save the toolbar settings)

## Expected behaviour

I can configure the toolbar position and it is saved (or restored)  well.

## Steps to reproduce

1. Get a dev version of GE (portable from CI or your custom version) and launch it
2. Change toolbar setup
3. Restart
4. See toolbar position configuration is the same as initally

## Screenshots

See above.

## Did this work in previous version of GitExtensions

Never since the feature was introduced... 

## Environment

- Git Extensions 33.33.33
- Build 3245408651bd658ee636e7c3b57f3a02bbe4abf4
- Git 2.28.0.windows.1
- Microsoft Windows NT 10.0.17134.0
- .NET Framework 4.8.4261.0
- DPI 192dpi (200% scaling)

## Diagnostics

The main goal to opening this issue is:

How can I debug it easily?
Where the configuration data are stored?

### PR DESCRIPTION
**Pull request by RussKie**
Due to discovered layout and placement instability of toolbars
reverting most of the functionality added in #8572.

Closes #8680

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Review by gerhardol — APPROVED** [review]
works for me

**Comment by pmiossec** [comment]
In my case, that doesn't solve the main problem I introduced with #8572, that is that the "Branch/filter" toolbar is displayed before the "Main" toolbar:

image

So, with this PR, this is somewhat worse because this layout is now no more changeable...

@RussKie Do you know where the toolbar positions are saved? I would like to see if the values are well saved...

**Comment by RussKie** [comment]
> Do you know where the toolbar positions are saved? I would like to see if the values are well saved...

I don't think this has anything to do with the saved data, as it is no longer used - I've removed all associated code.
This may be an artifact of changes in the designer though

**Comment by RussKie** [comment]
@pmiossec I've run this on two different PCs and haven't observed any issues.
Are you having issues in a dev version launched from VS? Can you try run `git clean -xdf` and then `.\build.cmd -launch` and see if you still get issues?
The settings are stored in a local user config located somewhere like this: C:\Users\<user>\AppData\Local\Git_Extensions\GitExtensions.exe_Url_<some crazy string>\33.33.33.0\user.config but I don't believe it has any relevance now, as we no longer persist/load the data from it.

**Comment by pmiossec** [comment]
>Are you having issues in a dev version launched from VS? Can you try run git clean -xdf and then .\build.cmd -launch and see if you still get issues?

Created a new worktree (so clean workspace) and I have this behavior (still the same):

bug_toolbar

>The settings are stored in a local user config located somewhere like this: C:\Users\<user>\AppData\Local\Git_Extensions\GitExtensions.exe_Url_<some crazy string>\33.33.33.0\user.config but I don't believe it has any relevance now, as we no longer persist/load the data from it.

I will still have a look...

**Comment by pmiossec** [comment]
>The settings are stored in a local user config located somewhere like this: C:\Users<user>\AppData\Local\Git_Extensions\GitExtensions.exe_Url_\33.33.33.0\user.config but I don't believe it has any relevance now, as we no longer persist/load the data from it.

What didn't work:
* removing this `user.config` file
* removing the `WindowPositions.xml` file

But changing the DPI scale to 100% instead of 200% fix the toolbar positioning.
Another DPI scale problem 😭

**Comment by RussKie** [comment]
So any latest build renders correctly on <200%, and incorrect on 200%?

**Comment by pmiossec** [comment]
> So any latest build renders correctly on <200%, and incorrect on 200%?

No, it took me some time to find the reason, so I just had some time to test 100% and 200% (that I use every days) with the code corresponding to this PR. 

I can just assert that every build since  the introduction of the 2 toolbars (with this revert or not) display badly 200%.

So before merging this revert, we have to test different dpi scaling to better understand. 

And if doesn't work, perhaps we will just have to revert the cut in 2 toolbars (until we find a fix)

**Comment by pmiossec** [comment]
Quick test: only 100% display toolbars well. 125%, 150%, 175%, 200% and 225% failed (and I stopped here)

**Comment by RussKie** [comment]
I can repro the issue, thank you. Investigating

**Comment by RussKie** [comment]
Pushed a possible workaround, please try it.

**Comment by RussKie** [comment]
I can't seem to repro the issue in a standalone repro, so quite likely the origin of the bug is same as for https://github.com/gitextensions/gitextensions/pull/8572 and https://github.com/gitextensions/gitextensions/issues/8698.

**Comment by RussKie** [comment]
It took me awhile, but I got a clean repro: https://github.com/RussKie/Test-ToolStripOrder

**Comment by RussKie** [comment]
@pmiossec have had a chance to test?

**Comment by pmiossec** [comment]
>It took me awhile, but I got a clean repro: https://github.com/RussKie/Test-ToolStripOrder
> @pmiossec have had a chance to test?

Yes, I reproduce with this repository

**Comment by RussKie** [comment]
Were you able to test the latest fix?

**Comment by pmiossec** [comment]
> 
> Were you able to test the latest fix?

Which fix?

If that's the 2nd commit (f362cd397fea0972624765e95118222ba9edcb25), after a `msbuild` and run, that doesn't work. Result:

image

**Comment by RussKie** [comment]
> Which fix?
> 
> If that's the 2nd commit (f362cd397fea0972624765e95118222ba9edcb25), after a `msbuild` and run, that doesn't work. Result:

There are only 2 commits in this PR, which one were you testing?
image

I've tested the fix at 150% and 200% without any issues.

**Review by pmiossec — APPROVED** [review]
After a new test, work as expected.

**Comment by pmiossec** [comment]
2 things about the cutting of the toolbars.

1. I find it strange to have the user scripts, that are added to the "Standard" toolbar, to be displayed before the "Filters" toolbar, like that:

image

I think we should create a new toolbar and display it after the 2 other toolbars to get something like that (so that we kept the toolbar layout in the current release):
image


2. I don't really understand why this PR revert also the hiding of the toolbar that works well (now that we discovered what was the cause of the toolbar bad placement and that you provided a workaround).
As you kept the cutting of the toolbar in 2 (i.e. that you didn't revert the whole development), perhaps we could at least kept one feature allowed by this cutting, no!?!

Please have a look at my branch that did this 2 things:

https://github.com/pmiossec/gitextensions/tree/new_toolbar_for_scripts

**Comment by RussKie** [comment]
> Please have a look at my branch that did this 2 things:
> 
> [https://github.com/pmiossec/gitextensions/tree/new_toolbar_for_scripts](https://github.com/pmiossec/gitextensions/tree/new_toolbar_for_scripts?rgh-link-date=2021-01-17T00%3A08%3A19Z)

Thank you, works for me :)
Though the new toolbar must look the same as the other two:
image

File: GitUI/CommandsDialogs/FormBrowse.cs [hunk 1]
```diff
@@ -312,6 +314,17 @@ public FormBrowse([NotNull] GitUICommands commands, string filter, ObjectId sele
 
             return;
 
+            void WorkaroundToolbarLocationBug()
+            {
+                // Layout engine bug (?) which may change the order of toolbars
+                // if the 1st one becomes longer than the 2nd toolbar's Location.X
+                // the layout engine will be place the 2nd toolbar first
+                toolPanel.TopToolStripPanel.Controls.Clear();
+                toolPanel.TopToolStripPanel.Controls.Add(ToolStripScripts);
```
Comment (gerhardol) [inline GitUI/CommandsDialogs/FormBrowse.cs:323]: If the order is left-to-right display order, should the order consistently be Filters, Sscripts, Main?

and I would prefer to remove/add button by button instead, splitting the toolbar is not adding much...

**Review by gerhardol — COMMENTED** [review]
+1

File: GitUI/CommandsDialogs/FormBrowse.cs (hunk 1 above)
Comment (RussKie) [reply GitUI/CommandsDialogs/FormBrowse.cs:323]: > splitting the toolbar is not adding much

It is adding if a) toolbar can be moved, and b) toolbars can be shown/hidden. "a" obviously doesn't work at this stage, but "b" is working just fine.

Comment (RussKie) [reply GitUI/CommandsDialogs/FormBrowse.cs:323]: > should the order consistently be Filters, Sscripts, Main?

The order is: Main, Filters, Scripts.

**Comment by RussKie** [comment]
> 1. I don't really understand why this PR revert also the hiding of the toolbar that works well (now that we discovered what was the cause of the toolbar bad placement and that you provided a workaround).
>    As you kept the cutting of the toolbar in 2 (i.e. that you didn't revert the whole development), perhaps we could at least kept one feature allowed by this cutting, no!?!

We also need to restore persistence of shown/hidden state as well.

**Comment by RussKie** [comment]
Merging, we can deal with saving/restoring visibility of the toolbars separately.

**Pull request merged**